Mathlib/Combinatorics/Additive/Randomisation.lean

```diff
@@ -0,0 +1,50 @@
+/-
+Copyright (c) 2023 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import Mathlib.Analysis.Fourier.FiniteAbelian.Orthogonality
+import Mathlib.Combinatorics.Additive.Dissociation
+
+/-!
+# Randomising by a function of dissociated support
+
+This file proves that a function from a finite abelian group can be randomised by a function of
+dissociated support.
+
+Precisely, for `G` a finite abelian group and two functions `c : Dual G → ℝ` and `d : Dual G → ℝ`
+such that `{ψ | d ψ ≠ 0} is dissociated`, the product of the `c ψ` over `ψ` is the same as the
+average over `a` of the product of the `c ψ + Re (d ψ * ψ a)`.
+-/
+
+open Finset
+open scoped BigOperators ComplexConjugate
+
+variable {G : Type*} [Fintype G] [AddCommGroup G] {p : ℕ}
+
+/-- One can randomise by a function of dissociated support. -/
+lemma AddDissociated.randomisation (c : AddChar G ℂ → ℝ) (d : AddChar G ℂ → ℂ)
+    (hcd : AddDissociated {ψ | d ψ ≠ 0}) : 𝔼 a, ∏ ψ, (c ψ + (d ψ * ψ a).re) = ∏ ψ, c ψ := by
+  refine Complex.ofReal_injective ?_
+  push_cast
+  calc
+    _ = ∑ t, (𝔼 a, ∏ ψ ∈ t, ((d ψ * ψ a) + conj (d ψ * ψ a)) / 2) * ∏ ψ ∈ tᶜ, (c ψ : ℂ) := by
+        simp_rw [expect_mul, ← expect_sum_comm, ← Fintype.prod_add, add_comm,
+          Complex.re_eq_add_conj]
+    _ = (𝔼 a, ∏ ψ ∈ ∅, ((d ψ * ψ a) + conj (d ψ * ψ a)) / 2) * ∏ ψ ∈ ∅ᶜ, (c ψ : ℂ) :=
+        Fintype.sum_eq_single ∅ fun t ht ↦ mul_eq_zero_of_left ?_ _
+    _ = _ := by simp
```

Comment:
I'd say we should put the rhs in here explicitly, for readability and robustness reasons. If something gets added to `simp` here, then it doesn't seem particularly nice to figure out from the following six lines what's gone wrong

Reply:
Sure, but note that `simp` doesn't see the RHS as a metavariable here. The RHS is `∏ ψ : AddChar G ℂ, ↑(c ψ)`, as expected